include/ginkgo/core/config/property_tree.hpp

```diff
@@ -0,0 +1,210 @@
+     */
+    pnode(const array_type& array);
+
+    /**
+     * Constructor for map
+     *
+     * @param map  a (string, pnode)-map
+     */
+    pnode(const map_type& map);
+
+    /**
+     * bool conversion. It's true if and only if it is not empty.
+     */
+    operator bool() const noexcept;
+
+    /**
+     * Get the current node status.
+     *
+     * @return the status
+     */
+    status_t get_status() const;
+
+    /**
+     * Get the array. It will throw error if the current node does not hold an
+     * array.
+     *
+     * @return the array const reference
+     */
+    const array_type& get_array() const;
+
+    /**
+     * Get the map. It will throw error if the current node does not hold an
+     * map.
+     *
+     * @return the map const reference
+     */
+    const map_type& get_map() const;
+
+    /**
+     * Get the boolean value. It will throw error if the current node does not
+     * hold an boolean value.
+     *
+     * @return the boolean value
+     */
+    bool get_boolean() const;
+
+    /**
+     * Get the integer value. It will throw error if the current node does not
+     * hold an integer value.
+     *
+     * @return the integer value with type int64_t
+     */
+    std::int64_t get_integer() const;
+
+    /**
+     * Get the real floating point value. It will throw error if the current
+     * node does not hold an real value.
+     *
+     * @return the real value with type double
+     */
+    double get_real() const;
+
+    /**
+     * Get the string. It will throw error if the current node does not hold an
+     * string.
+     *
+     * @return the string
+     */
+    std::string get_string() const;
+
+    /**
+     * This function is to access the data under the map. It will throw error
+     * when it does not hold a map. When access non-existent key in the map, it
+     * will return an empty node.
+     *
+     * @param key  the key of the map
+     *
+     * @return node const reference. If the map does not have the key, return
+     * an empty node.
+     */
+    const pnode& get(const std::string& key) const;
+
+    /**
+     * This function is to access the data under the array. It will throw error
+     * when it does not hold an array or access out-of-bound index.
+     *
+     * @param index  the index for array
+     *
+     * @return node const reference.
+     */
+    const pnode& get(int index) const;
+
+private:
+    void throw_if_not_contain(status_t status) const;
+
+    static const pnode& empty_node();
+
+    status_t status_;
+    array_type array_;  // for array
+    map_type map_;      // for map
+    // value
+    std::string str_;
+    union {
+        std::int64_t integer_;
+        double real_;
+        bool boolean_;
+    } union_data_;
+};
+
+
+template <typename T, std::enable_if_t<std::is_integral<T>::value>*>
+pnode::pnode(T integer) : status_(status_t::integer)
+{
+    if (integer > std::numeric_limits<std::int64_t>::max()) {
```

Comment:
just to be safe, we should also check the negative end of the range

Reply:
done.
I original only considered the IEEE interger <= 64 bits, which means int64_t provides the largest negative number range.